src/drivers/src/virtual_spi.rs

```diff
@@ -0,0 +1,163 @@
+use common::{List, ListLink, ListNode};
+use common::take_cell::TakeCell;
+use core::cell::Cell;
+use hil;
+
+/// The Mux struct manages multiple SPI clients. Each client may have
+/// at most one outstanding SPI request.
+pub struct MuxSPIMaster<'a> {
+    spi: &'a hil::spi::SpiMaster,
+    devices: List<'a, SPIMasterDevice<'a>>,
+    inflight: TakeCell<&'a SPIMasterDevice<'a>>,
+}
+
+impl<'a> hil::spi::SpiMasterClient for MuxSPIMaster<'a> {
+    fn read_write_done(&self,
+                       write_buffer: &'static mut [u8],
+                       read_buffer: Option<&'static mut [u8]>,
+                       len: usize) {
+        self.inflight.take().map(move |device| {
+            device.read_write_done(write_buffer, read_buffer, len);
+        });
+        self.do_next_op();
+    }
+}
+
+impl<'a> MuxSPIMaster<'a> {
+    pub const fn new(spi: &'a hil::spi::SpiMaster) -> MuxSPIMaster<'a> {
+        MuxSPIMaster {
+            spi: spi,
+            devices: List::new(),
+            inflight: TakeCell::empty(),
+        }
+    }
+
+    fn do_next_op(&self) {
+        if self.inflight.is_none() {
+            let mnode = self.devices.iter().find(|node| node.operation.get() != Op::Idle);
+            mnode.map(|node| {
+
+                match node.operation.get() {
+                    Op::Configure(cpol, cpal, rate) => {
+
+                        match node.chip_select {
+                            Some(x) => {
+                                self.spi.set_chip_select(x);
+                            }
+                            None => {}
+                        }
+
+                        // In theory, the SPI interface should support
+                        // using a GPIO in lieu of a hardware CS line.
+                        // This is particularly important for the SAM4L
+                        // if using a USART, but might be relevant
+                        // for other platforms as well.
+                        // TODO: make this do something if given GPIO pin
+                        // match node.chip_select_gpio {
+                        //     Some() => { },
+                        //     None => {}
+                        // }
```

Comment:
Alternatively, it could just be internalized in the `SPIMaster` implementation, so the interface, from the perspective of the client, doesn't care if it's part of the controller or a separate GPIO. I think there might be a nice way to do this with an associated type on the `SpiMaster` trait, which we can experiment with when we do the USART implementation